### PR TITLE
DataFrame.pivot_table

### LINKED ISSUE #382: Implement missing functions for 10-minute tutorial with Koalas
This is the list of missing functions for **10-minute tutorial with Koalas** (#196) 

+ [ ] DataFrame.pivot_table
+ [ ] DataFrame.reindex
+ [ ] DataFrame.apply
+ [x] DataFrame.append (#388)
+ [x] DataFrame.at (#384)
+ [ ] DataFrame.T
+ [x] DataFrame.sub (#427)

Ignored functions
+ DataFrame.to_hdf
+ DataFrame.iat
+ DataFrame.shift

Some functions could be implemented for both DataFrame and Series. 

For contributors, please mention that you're working on some functions here and go ahead.

### PR DESCRIPTION
**Pull request by garawalid**
+ Initial implementation of `DataFrame.pivot_table()`
+ Ignored parameters : `margins`, `dropna` and `margins_name`
+ Resolves #382

### CODE REVIEW

File: databricks/koalas/frame.py [hunk 1]
```diff
@@ -1653,6 +1653,58 @@ def head(self, n=5):
 
         return DataFrame(self._sdf.limit(n), self._metadata.copy())
 
+    def pivot_table(self, values=None, index=None, columns=None,
+                    aggfunc='mean', fill_value=None):
+
+        if not isinstance(columns, str):
+            raise ValueError("columns should be string.")
```
Comment (floscha) [inline databricks/koalas/frame.py]: I know the code base is not 100% consistent in this regard, but error messages should probably start with uppercase letters.

File: databricks/koalas/frame.py [hunk 2]
```diff
@@ -1653,6 +1653,58 @@ def head(self, n=5):
...
+                isinstance(key, str) and isinstance(value, str) for key, value in aggfunc.items())):
+            raise ValueError("aggfunc must be a dict mapping from column name (string) "
+                             "to aggregate functions (string).")
+
+        if isinstance(aggfunc, dict) and index is None:
+            raise NotImplementedError("pivot_table doesn't support aggfuct"
+                                      " as dict and without index.")
+
+        if isinstance(aggfunc, str):
+
```
Comment (floscha) [inline databricks/koalas/frame.py]: This empty line should be removed.

File: databricks/koalas/tests/test_dataframe.py [hunk 3]
```diff
@@ -569,3 +569,45 @@ def test_sample(self):
...
+
+        msg = "values should be string or list of columns."
+        with self.assertRaisesRegex(ValueError, msg):
+            kdf.pivot_table(index=['c'], columns="a", values=5)
+
+        msg = "index should be a None or a list of columns."
+        with self.assertRaisesRegex(ValueError, msg):
+            kdf.pivot_table(index="c", columns="a", values="b")
+
+        pdf = pd.DataFrame({'a': [4, 2, 3, 4, 8, 6],
```
Comment (floscha) [inline databricks/koalas/tests/test_dataframe.py]: Could this DataFrame not simply be defined in line 575?

**Comment by floscha** [comment]
The unit tests you've added so far only assert exception handling. It would be great if you could add both docstest and unit tests that assert the actual pivoting functionality. You could use the [pandas examples](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.pivot_table.html) to start with.

**Comment by garawalid** [comment]
@floscha thanks for your comments. I skipped the unit test because the result is not exactly the same and requires changing the design of the metadata. I'll add some examples in the docstring. 

Here is an example : 
```python 
print(kdf.pivot_table(index = ['c'], columns="a",values="b"))
print(pdf.pivot_table(index = ['c'], columns=["a"],values="b"))

print(kdf.pivot_table(index = ['c'], columns="a",values = ['b','e'],aggfunc={'b':'mean','e':'sum'}))
print(pdf.pivot_table(index = ['c'], columns=["a"],values = ['b','e'], aggfunc={'b':'mean','e':'sum'}))
```

```
     2    3    4    6    8
c                         
7  NaN  NaN  NaN  NaN  2.0
9  NaN  2.0  NaN  NaN  NaN
1  NaN  NaN  1.0  NaN  NaN
2  2.0  NaN  NaN  NaN  NaN
4  NaN  NaN  4.0  4.0  NaN
  
a    2    3    4    6    8
c                         
1  NaN  NaN  1.0  NaN  NaN
2  2.0  NaN  NaN  NaN  NaN
4  NaN  NaN  4.0  4.0  NaN
7  NaN  NaN  NaN  NaN  2.0
9  NaN  2.0  NaN  NaN  NaN


   2_b  2_e  3_b  3_e  4_b  4_e  6_b  6_e  8_b  8_e
c                                                  
7  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  2.0  2.0
9  NaN  NaN  2.0  2.0  NaN  NaN  NaN  NaN  NaN  NaN
1  NaN  NaN  NaN  NaN  1.0  1.0  NaN  NaN  NaN  NaN
2  2.0  2.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
4  NaN  NaN  NaN  NaN  4.0  4.0  4.0  4.0  NaN  NaN
  
     b                        e                    
a    2    3    4    6    8    2    3    4    6    8
c                                                  
1  NaN  NaN  1.0  NaN  NaN  NaN  NaN  1.0  NaN  NaN
2  2.0  NaN  NaN  NaN  NaN  2.0  NaN  NaN  NaN  NaN
4  NaN  NaN  4.0  4.0  NaN  NaN  NaN  4.0  4.0  NaN
7  NaN  NaN  NaN  NaN  2.0  NaN  NaN  NaN  NaN  2.0
9  NaN  2.0  NaN  NaN  NaN  NaN  2.0  NaN  NaN  NaN

```

File: databricks/koalas/tests/test_dataframe.py (hunk 3 above)
Comment (garawalid) [reply databricks/koalas/tests/test_dataframe.py]: I checked it and it works. Thanks!

File: databricks/koalas/tests/test_dataframe.py [hunk 4]
```diff
@@ -569,3 +569,45 @@ def test_sample(self):
             kdf.sample()
         with self.assertRaises(NotImplementedError):
             kdf.sample(n=1)
+
+    def test_pivot_table(self):
```
Comment (thunterdb) [inline databricks/koalas/tests/test_dataframe.py]: you should be able to replicate the dask test in https://github.com/dask/dask/blob/master/dask/dataframe/tests/test_reshape.py#L188

(without the categorical example)

Comment (garawalid) [reply databricks/koalas/tests/test_dataframe.py]: Thanks @thunterdb. 
I kept some tests as `Todo`, I'll update them once we implement hierarchical columns and column's name.

Comment (thunterdb) [reply databricks/koalas/tests/test_dataframe.py]: Sounds great.

**Comment by floscha** [comment]
I see your point @garawalid and I agree that changing Koalas' way of handling multilevel indices/columns should not be part of this PR. However, pandas' `pivot_table` appears to sort the indices which can also easily be achieved with Koalas by calling `.sort_index()` on the pivoted DataFrame. Maybe you should add this to better conform with the expectations of pandas users. On the downside, `.sort_index()` might cause some computational overhead that I cannot really estimate.

**Comment by garawalid** [comment]
Thanks @floscha for pointing out `sort_index()`, I think we can leave it like this and Pandas users will do it manually if necessary. In fact, we have to keep in mind that the DataFrame can be big, and they won't be able to see the full DataFrame. Also, the cost of sorting becomes an overhead if the DataFrame is big. Meanwhile, we can use `sort_index()` when we compare results in tests

File: databricks/koalas/tests/test_dataframe.py (hunk 4 above)
Comment (garawalid) [reply databricks/koalas/tests/test_dataframe.py]: Good.

File: databricks/koalas/frame.py [hunk 5]
```diff
@@ -1653,6 +1652,141 @@ def head(self, n=5):
...
 
+    def pivot_table(self, values=None, index=None, columns=None,
+                    aggfunc='mean', fill_value=None):
+        """
+        Create a spreadsheet-style pivot table as a DataFrame. The levels in
+        the pivot table will be stored in MultiIndex objects (hierarchical
+        indexes) on the index and columns of the result DataFrame.
+
+        Parameters
+        ----------%s
```
Comment (HyukjinKwon) [inline databricks/koalas/frame.py]: seems `%s` should be removed.

**Comment by HyukjinKwon** [comment]
hey @ganeshchand can we update this one? It's nice to have it.

**Comment by floscha** [comment]
Didn't you mean to ping @garawalid?

**Comment by HyukjinKwon** [comment]
D'oh

File: databricks/koalas/frame.py (hunk 5 above)
Comment (garawalid) [reply databricks/koalas/frame.py]: I'll take care of it.

**Comment by garawalid** [comment]
@HyukjinKwon thanks for the review. I'll update this PR. I've been quite busy lately, I'll join you ASAP. 

@HyukjinKwon @floscha 
I've been thinking about discarding the case where `values` is a list. And, we can update it later once we have multi level columns. 

What do you think ?

**Comment by floscha** [comment]
I think generally it's fine to only support one value for now and add multi-level support later. However, I would still treat `values` as a list (for type hints and stuff) but raise a `ValueError` for `len(values) > 1`, telling the user that only pivoting for one level is available for now. Then someone can add multi-level support later without changing the signature of the method.

**Comment by garawalid** [comment]
@floscha  thanks for your suggestion. 

I disabled multi level columns for instance. 
@floscha, @HyukjinKwon  would you mind review this PR.

File: databricks/koalas/frame.py [hunk 6]
```diff
@@ -2467,6 +2467,145 @@ def head(self, n=5):
...
```
Comment (HyukjinKwon) [inline databricks/koalas/frame.py]: Let's remove `%s`. Otherwise, I think @floscha's suggestion sounds fine.

File: databricks/koalas/frame.py [hunk 7]
```diff
@@ -2467,6 +2467,145 @@ def head(self, n=5):
...
+        A   B
+        foo one    4.0      1
+            two    NaN      6
+        bar two    7.0      6
+            one    4.0      5
+
+        We can also fill missing values using the `fill_value` parameter.
+
+        >>> table = df.pivot_table(values='D', index=['A', 'B'],
+        ...                     columns='C', aggfunc='sum', fill_value=0)
```
Comment (HyukjinKwon) [inline databricks/koalas/frame.py]: looks indentation should be fixed

File: databricks/koalas/frame.py [hunk 8]
```diff
@@ -2467,6 +2467,145 @@ def head(self, n=5):
...
+        5  bar  one  large  4  6
+        6  bar  one  small  5  8
+        7  bar  two  small  6  9
+        8  bar  two  large  7  9
+
+        This first example aggregates values by taking the sum.
+
+        >>> table = df.pivot_table(values='D', index=['A', 'B'],
+        ...                     columns='C', aggfunc='sum')
+        >>> table # doctest: +SKIP
```
Comment (HyukjinKwon) [inline databricks/koalas/frame.py]: two spaces before inlined comment (per pep8)

Comment (HyukjinKwon) [reply databricks/koalas/frame.py]: Actually, why do we need to skip? I think we can do `# doctest: +NORMALIZE_WHITESPACE` instead.

**Comment by garawalid** [comment]
@HyukjinKwon thanks for your review. Of course, I'll open it.

File: databricks/koalas/frame.py (hunk 8 above)
Comment (garawalid) [reply databricks/koalas/frame.py]: It works like a charm, thanks! 
Btw, do you use some plug-in in PyCharm to check pep8 style inside the docstring ?

File: databricks/koalas/frame.py (hunk 7 above)
Comment (garawalid) [reply databricks/koalas/frame.py]: Still not sure if indentation is fixed. @HyukjinKwon would you mind please to check it again.

**Comment by HyukjinKwon** [comment]
Thanks for working on this @garawalid. this is awesome.

Merged

**Pull request merged**